### PR TITLE
fix/decimal on arm

### PR DESCRIPTION
**Pull request by rickif**
This PR solve following test failures on aarch64:
```
FAILED TESTS (14/2318):
AggregateTest.test_avg
VectorizedDecimalCastExprFloatTest744.testCastFromFloatToDecimal32p9s2Abnormal
VectorizedDecimalCastExprFloatTest792.testCastFromFloatToDecimal32p7s4Abnormal
VectorizedDecimalCastExprFloatTest772.testCastFromFloatToDecimal32p9s9Abnormal
VectorizedDecimalCastExprFloatTest1122.testCastFromDoubleToDecimal32p9s9Abnormal
VectorizedDecimalCastExprFloatTest1142.testCastFromDoubleToDecimal32p7s4Abnormal
VectorizedDecimalCastExprFloatTest1076.testCastFromDoubleToDecimal32p9s0Abnormal
VectorizedDecimalCastExprFloatTest901.testCastFromFloatToDecimal64p15s13Abnormal
VectorizedDecimalCastExprFloatTest1239.testCastFromDoubleToDecimal64p18s18Abnormal
VectorizedDecimalCastExprFloatTest1267.testCastFromDoubleToDecimal64p15s13Abnormal
VectorizedDecimalCastExprFloatTest1100.testCastFromDoubleToDecimal32p9s2Abnormal
VectorizedDecimalCastExprFloatTest1212.testCastFromDoubleToDecimal64p18s9Abnormal
VectorizedDecimalCastExprFloatTest880.testCastFromFloatToDecimal64p18s18Abnormal
DecimalV2ValueTest.div 
```

### CODE REVIEW

**Review by kangkaisen — DISMISSED** [review]

File: be/src/runtime/decimalv3.h [hunk 1]
```diff
@@ -189,13 +189,19 @@ class DecimalV3Cast {
...
+    template <typename T>
+    static constexpr T float_upper_overflow_indicator = std::numeric_limits<T>::min();
+
     template <typename From, typename To>
     static inline bool from_float(FloatType<From> value, DecimalType<To> const& scale_factor,
                                   DecimalType<To>* dec_value) {
         *dec_value = static_cast<To>(scale_factor * static_cast<double>(value));
         if constexpr (is_decimal32<To> || is_decimal64<To>) {
-            return *dec_value == float_overflow_indicator<To>;
+            // Depending on the compiler implement, std::numeric_limits<T>::max() or std::numeric_limits<T>::max() both could be returned,
```
Comment (chaoyli) [inline be/src/runtime/decimalv3.h:201]: Add some concise comment about different architecture.

Comment (rickif) [reply be/src/runtime/decimalv3.h:201]: Detail about the difference of these arch is added.

File: be/src/runtime/decimalv3.h [hunk 2]
```diff
@@ -189,13 +189,25 @@ class DecimalV3Cast {
...
+            // when overflow is happenning in casting.
+
+            // With GCC-10.3.0, the cast on aarch64 uses fcvtzs instruction, behaving as "carries all overflows to the output precision’s largest finite number with the sign of the result before rounding".
+            // (https://developer.arm.com/documentation/ddi0487/latest)
+
+            // Meanwhile, the cast on x86_64 uses cvttsd2siq instruction, behaving as "the indefinite integer value (80000000H) is returned".
+            // (https://www.felixcloutier.com/x86/cvttsd2si)
+            return (*dec_value == float_lower_overflow_indicator<To>) ||
```
Comment (sduzh) [inline be/src/runtime/decimalv3.h:209]: How about replacing the logical or `||` with the bit or `|`?   Will it give us a better performance?

Comment (rickif) [reply be/src/runtime/decimalv3.h:209]:
```
static DecimalType<T> float_overflow_indicator = static_cast<To>(std::numeric_limits<double>::max());
```
From the performance perspective, would it be better to define a `float_overflow_indicator` like this ?  And we could just simplify the code as bellow.
```
return *dec_value == float_overflow_indicator<To>;
```

Comment (sduzh) [reply be/src/runtime/decimalv3.h:209]: Sounds reasonable

File: be/src/runtime/decimalv3.h [hunk 3]
```diff
@@ -189,13 +189,25 @@ class DecimalV3Cast {
     }
 
     template <typename T>
-    static constexpr T float_overflow_indicator = std::numeric_limits<T>::lowest();
+    static constexpr T float_lower_overflow_indicator = std::numeric_limits<T>::max();
```
Comment (chaoyli) [inline be/src/runtime/decimalv3.h:192]: for double and float, std::numeric_limits<double>::lowest() is the actual minimum.
https://en.cppreference.com/w/cpp/types/numeric_limits

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: be/src/runtime/decimalv3.h (hunk 3 above)
Comment (rickif) [reply be/src/runtime/decimalv3.h:192]: The actual type of decimal, which we compare with here, is integer.

File: be/src/runtime/decimalv3.h (hunk 2 above)
Comment (rickif) [reply be/src/runtime/decimalv3.h]: I have done some simple benchmark about performance of `|` and `||`. It seems that there's no significant difference here. Hence, I think logical or is better here. @sduzh

File: be/src/runtime/decimalv3.h [hunk 4]
```diff
@@ -188,21 +189,22 @@ class DecimalV3Cast {
...
-        }
+        static_assert(is_decimal<To>, "invalid decimal type");
+
+        auto double_dec_value = static_cast<double>(value) * scale_factor;
+        *dec_value = static_cast<To>(double_dec_value);
+
+        // Find that exponent exp such that
+        //     y == x * 2^exp
+        // where x in (-1;-0.5],[0.5, 1.0), which implies that y <= 2^exp.
+        // If exp <= (sizeof(T) - 1), the y must be represnetable by type T.
```
Comment (chaoyli) [inline be/src/runtime/decimalv3.h:203]: spelling error, should be representable

File: be/src/runtime/decimalv3.h [hunk 5]
```diff
@@ -188,21 +189,22 @@ class DecimalV3Cast {
...
-            static_assert(is_decimal<To>, "invalid decimal type");
-        }
+        static_assert(is_decimal<To>, "invalid decimal type");
+
+        auto double_dec_value = static_cast<double>(value) * scale_factor;
+        *dec_value = static_cast<To>(double_dec_value);
+
+        // Find that exponent exp such that
+        //     y == x * 2^exp
+        // where x in (-1;-0.5],[0.5, 1.0), which implies that y <= 2^exp.
```
Comment (chaoyli) [inline be/src/runtime/decimalv3.h:202]: (-1, 0.5]

**Review by chaoyli — CHANGES_REQUESTED** [review]

**Review by satanson — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Pull request merged**